lib/auth/grpcserver.go

```diff
@@ -6570,7 +6571,37 @@ type grpcContext struct {
 func (g *GRPCServer) authenticate(ctx context.Context) (*grpcContext, error) {
 	// HTTPS server expects auth context to be set by the auth middleware
 	authContext, err := g.Authorizer.Authorize(ctx)
-	if err != nil {
+
+	// If authorization fails (e.g., IP pinning mismatch), we capture the rejection
+	// at the connection layer to ensure security visibility in the Audit Log/UI.
+	if err != nil {
+		p, _ := peer.FromContext(ctx)
+		identity, _ := authz.UserFromContext(ctx)
+
+		if p != nil && identity != nil {
+			event := &apievents.SessionReject{
+				Metadata: apievents.Metadata{
+					Type: events.SessionRejectedEvent,
+					Code: events.SessionRejectedCode,
+				},
+				UserMetadata: apievents.UserMetadata{
+					User: identity.GetIdentity().Username,
+				},
+				ConnectionMetadata: apievents.ConnectionMetadata{
+					RemoteAddr: p.Addr.String(),
+					Protocol:   events.EventProtocol,
+				},
+				ServerMetadata: apievents.ServerMetadata{
+					ServerVersion:   teleport.Version,
+					ServerID:        g.AuthServer.ServerID,
+					ServerNamespace: apidefaults.Namespace,
+				},
+				Reason: trace.Unwrap(err).Error(),
```

Comment:
Sometimes these trace errors print a pretty ugly stack trace message (it's fine for logs, but not so much for user-facing messages).

I wonder if this helps?
```suggestion
				Reason: trace.UserMessage(err),
```

We'll then want to look at `g.Authorizer.Authorize()` and make sure all of the authorization failures it produces have a user-friendly message.